testing.go

```diff
@@ -41,29 +54,108 @@ func StartTest(t *testing.T) *Test {
 	})
 	span.SetBaggageItem("trace.kind", "test")
 
-	return &Test{
+	// Replaces stdout and stderr
+	stdOut := newStdIO(&os.Stdout)
+	stdErr := newStdIO(&os.Stderr)
+	log2.SetOutput(stdOut.writePipe)
+
+	test := &Test{
 		ctx:  ctx,
 		span: span,
 		t:    t,
+		stdOut: stdOut,
+		stdErr: stdErr,
+	}
+
+	// Starts stdIO pipe handlers
+	if test.stdOut != nil {
+		go stdIOHandler(test, test.stdOut, false)
 	}
+	if test.stdErr != nil {
+		go stdIOHandler(test, test.stdErr, true)
+	}
+
+	return test
 }
 
 func (test *Test) End() {
+	defer test.span.Finish()
+
 	if r := recover(); r != nil {
 		test.span.SetTag("test.status", "ERROR")
+		test.stdOut.restore(&os.Stdout)
+		test.stdErr.restore(&os.Stderr)
+		test.span.Finish()
 		_ = GlobalAgent.Flush()
 		panic(r)
 	}
-	test.span.Finish()
+
 	if test.t.Failed() {
 		test.span.SetTag("test.status", "FAIL")
 	} else if test.t.Skipped() {
 		test.span.SetTag("test.status", "SKIP")
 	} else {
 		test.span.SetTag("test.status", "PASS")
 	}
+
+	test.stdOut.restore(&os.Stdout)
+	test.stdErr.restore(&os.Stderr)
+	log2.SetOutput(os.Stderr)
 }
 
 func (test *Test) Context() context.Context {
 	return test.ctx
 }
+
+// Handles the StdIO pipe for stdout and stderr
+func stdIOHandler (test *Test, stdio *StdIO, isError bool) {
+	stdio.sync.Add(1)
+	defer stdio.sync.Done()
+	reader := bufio.NewReader(stdio.readPipe)
```

Comment:
Did you fix the issue with the scanner using this?

Reply:
yes, I've replaced the scanner with the reader to reduce overhead in the pipe read and force a `*File.Sync()` on both pipes in the restore method. With that I didn't lost any stdio write again.